compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -712,50 +719,72 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(implicit ctx: Context) {
           }
         }
 
-        pat match {
-          case Typed(pat1, tpt) =>
-            val getBoundVars = new TreeAccumulator[List[TypeSymbol]] {
-              def apply(syms: List[TypeSymbol], t: Tree)(implicit ctx: Context) = {
-                val syms1 = t match {
-                  case t: Bind if t.symbol.isType =>
-                    t.symbol.asType :: syms
-                  case _ =>
-                    syms
-                }
-                foldOver(syms1, t)
+        def getBoundVarsList(pat: Tree, tpt: Tree): List[(TypeSymbol, Boolean)] = {
+          // UnApply nodes with pattern bound variables translate to something like this
+          //   UnApply[t @ t](pats)(implicits): T[t]
+          // Need to traverse any binds in type arguments of the UnAppyl to get the set of
+          // all instantiable type variables. Test case is pos/inline-caseclass.scala.
+          val allTpts = tpt :: (pat match {
+            case UnApply(TypeApply(_, tpts), _, _) => tpts
+            case _ => Nil
+          })
+
+          val getBinds = new TreeAccumulator[Set[TypeSymbol]] {
+            def apply(syms: Set[TypeSymbol], t: Tree)(implicit ctx: Context): Set[TypeSymbol] = {
+              val syms1 = t match {
+                case t: Bind if t.symbol.isType =>
+                  syms + t.symbol.asType
+                case _ => syms
               }
+              foldOver(syms1, t)
             }
-            var boundVars = getBoundVars(Nil, tpt)
-            // UnApply nodes with pattern bound variables translate to something like this
-            //   UnApply[t @ t](pats)(implicits): T[t]
-            // Need to traverse any binds in type arguments of the UnAppyl to get the set of
-            // all instantiable type variables. Test case is pos/inline-caseclass.scala.
-            pat1 match {
-              case UnApply(TypeApply(_, tpts), _, _) =>
-                for (tpt <- tpts) boundVars = getBoundVars(boundVars, tpt)
-              case _ =>
-            }
-            for (bv <- boundVars) {
-              val TypeBounds(lo, hi) = bv.info.bounds
-              ctx.gadt.addBound(bv, lo, isUpper = false)
-              ctx.gadt.addBound(bv, hi, isUpper = true)
+          }
+          val binds = allTpts.foldLeft[Set[TypeSymbol]](Set.empty)(getBinds.apply(_, _))
```

Comment:
Use this `apply` method in TreeAccumulator instead?

      def apply(x: X, trees: Traversable[Tree])(implicit ctx: Context): X = (x /: trees)(apply)